contracts/exchange-forwarder/test/utils/forwarder_test_factory.ts

```diff
@@ -0,0 +1,396 @@
+        makerAssetAmount = BigNumber.max(makerAssetAmount.minus(makerAssetFilled), constants.ZERO_AMOUNT);
+
+        currentState.takerAssetFillAmount = currentState.takerAssetFillAmount.plus(takerAssetAmount);
+        currentState.makerAssetFillAmount = currentState.makerAssetFillAmount.plus(makerAssetAmount);
+
+        if (order.takerFeeAssetData === order.makerAssetData) {
+            currentState.percentageFees = currentState.percentageFees.plus(takerFee);
+        } else if (order.takerFeeAssetData === order.takerAssetData) {
+            currentState.wethFees = currentState.wethFees.plus(takerFee);
+        }
+
+        remainingOrdersToFill = BigNumber.max(remainingOrdersToFill.minus(new BigNumber(1)), constants.ZERO_AMOUNT);
+    }
+
+    return currentState;
+}
+
+// Since bignumber is not compatible with chai's within
+function expectBalanceWithin(balance: BigNumber, low: BigNumber, high: BigNumber): void {
+    expect(balance).to.be.bignumber.gte(low);
+    expect(balance).to.be.bignumber.lte(high);
+}
+
+export class ForwarderTestFactory {
+    private readonly _exchangeWrapper: ExchangeWrapper;
+    private readonly _forwarderWrapper: ForwarderWrapper;
+    private readonly _erc20Wrapper: ERC20Wrapper;
+    private readonly _forwarderAddress: string;
+    private readonly _makerAddress: string;
+    private readonly _takerAddress: string;
+    private readonly _orderFeeRecipientAddress: string;
+    private readonly _forwarderFeeRecipientAddress: string;
+    private readonly _wethAddress: string;
+    private readonly _gasPrice: BigNumber;
+
+    public static getPercentageOfValue(value: BigNumber, percentage: BigNumber): BigNumber {
+        const numerator = constants.PERCENTAGE_DENOMINATOR.times(percentage).dividedToIntegerBy(100);
+        const newValue = value.times(numerator).dividedToIntegerBy(constants.PERCENTAGE_DENOMINATOR);
+        return newValue;
+    }
+
+    constructor(
+        exchangeWrapper: ExchangeWrapper,
+        forwarderWrapper: ForwarderWrapper,
+        erc20Wrapper: ERC20Wrapper,
+        forwarderAddress: string,
+        makerAddress: string,
+        takerAddress: string,
+        orderFeeRecipientAddress: string,
+        forwarderFeeRecipientAddress: string,
+        wethAddress: string,
+        gasPrice: BigNumber,
+    ) {
+        this._exchangeWrapper = exchangeWrapper;
+        this._forwarderWrapper = forwarderWrapper;
+        this._erc20Wrapper = erc20Wrapper;
+        this._forwarderAddress = forwarderAddress;
+        this._makerAddress = makerAddress;
+        this._takerAddress = takerAddress;
+        this._orderFeeRecipientAddress = orderFeeRecipientAddress;
+        this._forwarderFeeRecipientAddress = forwarderFeeRecipientAddress;
+        this._wethAddress = wethAddress;
+        this._gasPrice = gasPrice;
+    }
+
+    public async marketBuyTestAsync(
+        orders: SignedOrder[],
+        fractionalNumberOfOrdersToFill: BigNumber,
```

Comment:
It kind of feels like a cognitive waste to make this a `BigNumber` since it seems to always be something that fits in a `number`.